src/main/scala/sbtassembly/Assembly.scala

```diff
@@ -1,394 +1,703 @@
-        if (ao.includeBin) Some(jar) else None
+    val filteredJars = timed(Level.Debug, "Filter jars") {
+      jars.flatMap {
+        case jar if excludedJars contains jar.data.asFile => None
+        case jar if isScalaLibraryFile(scalaLibraries, jar.data.asFile) =>
+          if (ao.includeScala) Some(jar) else None
+        case jar if externalDeps contains jar.data.asFile =>
+          if (ao.includeDependency) Some(jar) else None
+        case jar =>
+          if (ao.includeBin) Some(jar) else None
+      }
     }
-    val dirRules = shadeRules.filter(_.isApplicableToCompiling)
-    val dirsFiltered =
-      dirs.par flatMap {
-        case dir =>
-          if (ao.includeBin) Some(dir)
-          else None
-      } map { dir =>
-        val hash = sha1name(dir.data)
-        IO.write(tempDir / (hash + "_dir.dir"), dir.data.getCanonicalPath, IO.utf8, false)
-        val dest = tempDir / (hash + "_dir")
-        if (dest.exists) {
-          IO.delete(dest)
-        }
-        dest.mkdir()
-        IO.copyDirectory(dir.data, dest)
-        if (dirRules.nonEmpty) {
-          val mappings = ((dest ** (-DirectoryFilter)).get  pair relativeTo(dest)) map {
-            case (k, v) => k.toPath -> v
+
+    val shadeRules = ao.shadeRules
+    val clazzShader = shader(shadeRules.filter(_.isApplicableToCompiling))
+    val classByParentDir =
+      if (!ao.includeBin) Vector.empty[(File, File)]
+      else
+        dirs
+          .flatMap { dir =>
+            (dir.data ** (-DirectoryFilter)).get.map(dir.data -> _)
           }
-          Shader.shadeDirectory(dirRules, dest.toPath, mappings, ao.level == Level.Debug)
-        }
-        dest
-      }
-    val jarDirs =
-      (for(jar <- libsFiltered.par) yield {
-        val jarName = jar.data.asFile.getName
-        val jarRules = shadeRules
-          .filter(r => (r.isApplicableToAll ||
-            jar.metadata.get(moduleID.key)
-              .map(m => ModuleCoordinate(m.organization, m.name, m.revision))
-              .exists(r.isApplicableTo)))
-        val hash = sha1name(jar.data) + "_" + sha1content(jar.data) + "_" + sha1rules(jarRules)
-        val jarNamePath = tempDir / (hash + ".jarName")
-        val dest = tempDir / hash
-        // If the jar name path does not exist, or is not for this jar, unzip the jar
-        if (!ao.cacheUnzip || !jarNamePath.exists || IO.read(jarNamePath) != jar.data.getCanonicalPath )
-        {
-          log.debug("Including: %s".format(jarName))
-          IO.delete(dest)
-          dest.mkdir()
-          AssemblyUtils.unzip(jar.data, dest, log)
-          IO.delete(ao.excludedFiles(Seq(dest)))
-          if (jarRules.nonEmpty) {
-            val mappings = ((dest ** (-DirectoryFilter)).get pair relativeTo(dest)) map {
-              case (k, v) => k.toPath -> v
-            }
-            Shader.shadeDirectory(dirRules, dest.toPath, mappings, ao.level == Level.Debug)
+    val clazzMappings =
+      timed(Level.Info, "Collect project classes") {
```

Comment:
Looking at the CI log, it shows:

```
[info] [info] Collect project classes start:
[info] [info] Collect project classes end. Took 0 ms
[info] [info] Collect dependency entries start:
[info] [info] Collect dependency entries end. Took 84 ms
[info] [info] Collect merge strategies start:
[info] [info] Collect merge strategies end. Took 29 ms
[info] [info] Merge all conflicting jar entries start:
[info] [info] 1 file(s) merged using strategy 'Discard' (Run the task at debug level to see the details)
[info] [info] Merge all conflicting jar entries end. Took 2378 ms
[info] [info] Finding remaining conflicts that were not merged start:
[info] [info] Finding remaining conflicts that were not merged end. Took 42 ms
[info] [info] Create jar start:
[info] [info] Create jar end. Took 606 ms
[info] [info] Jar hash: a6f5029aa2f99bc14c0a0461a0363d513c509bed
[info] [success] Total time: 3 s, completed Apr 16, 2022, 5:04:14 PM
```

If the entire operation is going to be really fast, and there's nothing actionable from the plugin users' point of view, I think we should remove all info level logs.

Reply:
I will move all these timers levels to DEBUG